packages/concerto-core/lib/introspect/mapkeytype.js

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const { MetaModelNamespace } = require('@accordproject/concerto-metamodel');
+
+const Decorated = require('./decorated');
+const IllegalModelException = require('./illegalmodelexception');
+
+// Types needed for TypeScript generation.
+/* eslint-disable no-unused-vars */
+/* istanbul ignore next */
+if (global === undefined) {
+    const ModelFile = require('./modelfile');
+}
+
+/**
+ * MapKeyType defines a Key type of an MapDeclaration.
+ *
+ * @extends Decorated
+ * @see See {@link Decorated}
+ * @class
+ * @memberof module:concerto-core
+ */
+class MapKeyType extends Decorated {
+    /**
+     * Create an MapKeyType.
+     * @param {MapDeclaration} parent - The owner of this property
+     * @param {Object} ast - The AST created by the parser
+     * @param {ModelFile} modelFile - the ModelFile for the Map class
+     * @throws {IllegalModelException}
+     */
+    constructor(parent, ast) {
+        super(ast);
+        this.parent = parent;
+        this.type = null;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+        this.type = this.ast.name;
+    }
+
+    /**
+     * Semantic validation of the structure of this class.
+     *
+     * @throws {IllegalModelException}
+     * @protected
+     */
+    validate() {
+        const declaration = this.getModelFile().getAllDeclarations();
```

Comment:
I was confused by this. The goal is to check the type of the key... 

I think I'd try this as a switch/case to make it clearer. Something like:

switch(this.getKeyType()) {
   case 'String':
   case 'Boolean':
   case 'DateTime':
    // valid!
   break;
  default:
 {
    // resolveType using model manager (throws exception if type is missing)
   // if type is scalar and not in [string, boolean, date time] throw error
   // if type is concept and not identified throw error
  }
}

This looks like it doesn't handle the case of key being null. You should use `BaseModelManager.resolveType` to lookup complex types so that if the lookup fails we get a consistent error message.

Reply:
In the case of a key or value, they can never be null, otherwise the MapType cannot be constructed.